src/utils/media-devices-manager.js

```diff
@@ -191,34 +202,68 @@ export default class MediaDevicesManager {
     }
   }
 
-  async setupNewMediaStream() {
-    await this.fetchMediaDevices();
+  async stopMicShare() {
+    this.audioSystem.removeStreamFromOutboundAudio("microphone");
 
-    // we should definitely have an audioTrack at this point unless they denied mic access
-    if (this.audioTrack) {
-      const micDeviceId = this.micDeviceIdForMicLabel(this.micLabelForAudioTrack(this.audioTrack));
-      if (micDeviceId) {
-        this._store.update({ settings: { lastUsedMicDeviceId: micDeviceId } });
-        console.log(`Selected input device: ${this.micLabelForDeviceId(micDeviceId)}`);
+    this.audioTrack?.stop();
+    this.audioTrack = null;
+
+    this._scene.emit("action_mute");
+
+    NAF.connection.adapter.enableMicrophone(false);
+    await NAF.connection.adapter.setLocalMediaStream(this._mediaStream);
+  }
+
+  async startVideoShare(constraints, isDisplayMedia, success, error) {
+    let newStream;
+    let videoTrackAdded = false;
+
+    try {
+      if (isDisplayMedia) {
+        newStream = await navigator.mediaDevices.getDisplayMedia(constraints);
+      } else {
+        newStream = await navigator.mediaDevices.getUserMedia(constraints);
       }
-      this._scene.emit("local-media-stream-created");
-    } else {
-      console.log("No available audio tracks");
+
+      const videoTracks = newStream ? newStream.getVideoTracks() : [];
+      if (videoTracks.length > 0) {
+        videoTrackAdded = true;
+
+        newStream.getVideoTracks().forEach(track => {
+          // Ideally we would use track.contentHint but it seems to be read-only in Chrome so we just add a custom property
+          track["_hubs_contentHint"] = isDisplayMedia ? "share" : "camera";
+          track.addEventListener("ended", () => {
+            this._scene.emit("action_end_video_sharing");
+          });
+          this._mediaStream.addTrack(track);
+        });
+
+        if (newStream && newStream.getAudioTracks().length > 0) {
+          this.audioSystem.addStreamToOutboundAudio("screenshare", newStream);
```

Comment:
Not related to this PR exactly, but now that I understand MediaSoup better I think we should eventually rework the stuff being done here to mix in screen share audio into the users output audio stream. Its fairly easy to send this as just another audio producer. This gets rid of some of the hoop jumping we are doing here to mix it, and probides a better experience anyway since the desktop audio can actually come from the screen share instead of the avatar. I have some ideas of adding a "group" concept to setLocalMediaStream/getMediaStream. We should discuss it.

Reply:
Screen share audio only works for Chrome right now and only for tab sharing for this might be a limited feature but we can definitely do it if we are willing to.

Reply:
Ah, what I mean is we area already doing this in Chrome, we just mix it into your outgoing audio stream. Once we rework naf-dialog-adapter to support multiple outgoing streams (we likely will do this first for video so that you can use a webcam and screen share at the same time) then we can get rid of this code that mixes your desktop audio with your mic audio and instead send it on its own producer.

Reply:
Oh yep, get it. Yep makes sense that we keep them separated so we can manipulate them independently on the client side although screen share audio seems to be quite limited right now.